File: web/lib/security.ts
/**
 * Security utilities for input sanitization and XSS prevention
 */

/**
 * Escape HTML entities to prevent XSS
 */
export function escapeHtml(text: string): string {
  const htmlEntities: Record<string, string> = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#x27;',
    '/': '&#x2F;',
    '`': '&#x60;',
    '=': '&#x3D;'
  }
  return text.replace(/[&<>"'`=/]/g, char => htmlEntities[char] || char)
}

/**
 * Sanitize URL to prevent javascript: and data: URLs
 */
export function sanitizeUrl(url: string): string {
  const trimmed = url.trim().toLowerCase()
  if (
    trimmed.startsWith('javascript:') ||
    trimmed.startsWith('data:') ||
    trimmed.startsWith('vbscript:')
  ) {
    return '#'
  }
  return url
}

/**
 * Convert markdown to HTML with XSS protection
 * This function escapes user content while preserving safe markdown formatting
 */
export function convertMarkdownToHtmlSafe(markdown: string): string {
  // First escape all HTML in the input
  let html = escapeHtml(markdown)
  
  // Handle code blocks (preserve them from other transformations)
  const codeBlocks: string[] = []
  html = html.replace(/```(\w*)\n?([\s\S]*?)```/g, (_, lang, code) => {
    const placeholder = `__CODE_BLOCK_${codeBlocks.length}__`
    codeBlocks.push(`<pre class="bg-gray-800 text-gray-100 p-3 rounded-lg text-xs overflow-x-auto my-2"><code>${code.trim()}</code></pre>`)
    return placeholder
  })
  
  // Handle headers
  html = html.replace(/^### (.+)$/gm, '<h3 class="font-semibold text-sm mt-3 mb-1">$1</h3>')
  html = html.replace(/^## (.+)$/gm, '<h2 class="font-semibold text-base mt-3 mb-1">$1</h2>')
  html = html.replace(/^# (.+)$/gm, '<h1 class="font-bold text-lg mt-3 mb-2">$1</h1>')
  
  // Handle bold
  html = html.replace(/\*\*([^*\n]+)\*\*/g, '<strong>$1</strong>')
  // Handle italic
  html = html.replace(/(?<!\*)\*([^*\n]+)\*(?!\*)/g, '<em>$1</em>')
  
  // Handle inline code
  html = html.replace(/`([^`\n]+)`/g, '<code class="bg-gray-100 px-1 py-0.5 rounded text-xs">$1</code>')
  
  // Restore code blocks
  codeBlocks.forEach((block, i) => {
    html = html.replace(`__CODE_BLOCK_${i}__`, block)
  })
  
  // Handle links - sanitize URLs
  html = html.replace(/\[([^\]]+)\]\(([^)]+)\)/g, (_, text, url) => {
    const safeUrl = sanitizeUrl(url)
    return `<a href="${safeUrl}" class="text-emerald-600 hover:underline" target="_blank" rel="noopener noreferrer">${text}</a>`
  })
  
  // Process lists and paragraphs
  const lines = html.split('\n')
  const processed: string[] = []
  let inList = false
  let listType = ''
  
  for (const line of lines) {
    const trimmed = line.trim()
    if (!trimmed) {
      if (inList) { processed.push(listType === 'ul' ? '</ul>' : '</ol>'); inList = false }
      continue
    }
    
    const bulletMatch = trimmed.match(/^[-*]\s+(.+)$/)
    if (bulletMatch) {
      if (!inList || listType !== 'ul') {
        if (inList) processed.push(listType === 'ul' ? '</ul>' : '</ol>')
        processed.push('<ul class="list-disc list-inside my-2 space-y-1">')
        inList = true; listType = 'ul'
      }
      processed.push(`<li>${bulletMatch[1]}</li>`)
      continue
    }
    
    const numMatch = trimmed.match(/^\d+\.\s+(.+)$/)
    if (numMatch) {
      if (!inList || listType !== 'ol') {
        if (inList) processed.push(listType === 'ul' ? '</ul>' : '</ol>')
        processed.push('<ol class="list-decimal list-inside my-2 space-y-1">')
        inList = true; listType = 'ol'
      }
      processed.push(`<li>${numMatch[1]}</li>`)
      continue
    }
    
    if (inList) { processed.push(listType === 'ul' ? '</ul>' : '</ol>'); inList = false; listType = '' }
    // Allow safe HTML tags we created
    if (trimmed.startsWith('<')) { processed.push(trimmed); continue }
    processed.push(`<p class="my-1">${trimmed}</p>`)
  }
  if (inList) processed.push(listType === 'ul' ? '</ul>' : '</ol>')
  
  return processed.join('')
}

/**
 * Sanitize search input to prevent SQL injection in PostgREST queries
 */
export function sanitizeSearchInput(input: string): string {
  return input
    .replace(/[%_\\]/g, '') // Remove wildcards and backslash
    .replace(/[\x00-\x1f\x7f]/g, '') // Remove control characters
    .slice(0, 100) // Limit length
    .trim()
}

/**
 * Validate UUID format
 */
export function isValidUUID(str: string): boolean {
  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
  return uuidRegex.test(str)
}

/**
 * Sanitize filename for safe storage
 */
export function sanitizeFilename(filename: string): string {
  return filename
    .replace(/[^a-zA-Z0-9._-]/g, '_') // Replace unsafe chars with underscore
    .replace(/\.{2,}/g, '.') // Remove consecutive dots
    .slice(0, 255) // Limit length
}

File: server/pkg/security/security.go
package security

import (
	"regexp"
	"strings"
)

var htmlEntities = map[rune]string{
	'&':  "&amp;",
	'<':  "&lt;",
	'>':  "&gt;",
	'"':  "&quot;",
	'\'': "&#x27;",
	'/':  "&#x2F;",
	'`':  "&#x60;",
	'=':  "&#x3D;",
}

// EscapeHTML escapes HTML entities to prevent XSS.
func EscapeHTML(text string) string {
	var sb strings.Builder
	for _, r := range text {
		if esc, ok := htmlEntities[r]; ok {
			sb.WriteString(esc)
		} else {
			sb.WriteRune(r)
		}
	}
	return sb.String()
}

// SanitizeURL prevents javascript: and data: URLs.
func SanitizeURL(url string) string {
	trimmed := strings.ToLower(strings.TrimSpace(url))
	if strings.HasPrefix(trimmed, "javascript:") ||
		strings.HasPrefix(trimmed, "data:") ||
		strings.HasPrefix(trimmed, "vbscript:") {
		return "#"
	}
	return url
}

var uuidRegex = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`)

// IsValidUUID validates UUID format.
func IsValidUUID(str string) bool {
	return uuidRegex.MatchString(strings.ToLower(str))
}

var unsafeFilenameChars = regexp.MustCompile(`[^a-zA-Z0-9._-]`)
var consecutiveDots = regexp.MustCompile(`\.{2,}`)

// SanitizeFilename sanitizes a filename for safe storage.
func SanitizeFilename(filename string) string {
	safe := unsafeFilenameChars.ReplaceAllString(filename, "_")
	safe = consecutiveDots.ReplaceAllString(safe, ".")
	if len(safe) > 255 {
		safe = safe[:255]
	}
	return safe
}

// SanitizeSearchInput sanitizes search input to prevent SQL injection.
func SanitizeSearchInput(input string) string {
	// Remove wildcards and backslash
	result := strings.Map(func(r rune) rune {
		if r == '%' || r == '_' || r == '\\' {
			return -1
		}
		// Remove control characters
		if r < 0x20 || r == 0x7f {
			return -1
		}
		return r
	}, input)

	if len(result) > 100 {
		result = result[:100]
	}
	return strings.TrimSpace(result)
}
